### ISSUE TITLE
Tracking higher-order grads when arbitrarily combining submodules of functional module

### ISSUE THREAD

**Issue by dylandoblar** [facebookresearch/higher#106]
Hi, I'm implementing a bilevel optimization procedure to meta-learn an embedding and am struggling to find the best way to track higher-order gradients for all of my learned modules.  The model in my full pipeline involves several torch modules which are combined in different ways depending on where in the pipeline the forward pass happens.  I would like to meta-learn an embedding (optimized in the outer loop, possibly along with some of the model parameters), and optimize some model parameters in the inner loop (to produce "fast weights") where the inner loss is partially parameterized by this meta-learned embedding.

Currently, to fit into the `higher` framework, I'm wrapping all of the modules for my model (in addition to the learned embedding) in a super-module, with each module as a submodule.  I can then pass this super-module to a `higher.innerloop_ctx` to track higher order grads for all params of all of these modules.  Ideally, I'd like for each of the submodules to be functional as well, so that I can combine modules arbitrarily by referring to them as `fmodel.submodule1(...)` and have higher order gradients computed in the backward pass.  However, it seems like this is not supported by `higher`, and that the only way to do forward passes inside an `innerloop_ctx` is by calling the `fmodel`'s `forward` function.  This is undesirable in my use case, since it would require the `forward` function of the super-module to be very bloated with different conditions to account for all of the cases in which it could be called.

My questions are as follows:

1. Is there a way to make each of my submodules functional so that they can be combined arbitrarily and still track higher order gradients?
2. Do I need to be using this super-module at all?  Is it possible/simple to track higher order grads for several modules at once (allowing me to decouple my learned embedding from the rest of the model, for example)?

Below is a minimal runnable example to illustrate the issue I'm facing.  The inner and outer losses in the example are arbitrary and just provide _something_ to optimize.  The lines of interest are marked with `# NOTE:` comments.  The approaches using the `forward` method of `fmodel` (the super-module) produce the desired results, the others do not.

Essentially what I want is for `fmodel.decoder(fmodel.encoder(frame))` to behave the same way as `fmodel(frame)` in the example below — is this possible with `higher`?

```python
import torch
import torch.optim as optim
import torch.nn as nn

import higher

device = torch.device('cpu')
inner_lr = outer_lr = .001
channels = 3
batch_size = 7

# random dataloader for debugging
def batch_generator():
    while True:
        yield torch.rand((batch_size, channels, 64, 64), device=device)
dataloader = batch_generator()


class BasicModel(nn.Module):
    def __init__(self, nc, emb_dim):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Conv2d(nc, 16, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.Conv2d(16, 32, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.Conv2d(32, 64, 7)
        )
        self.decoder = nn.Sequential(
            nn.ConvTranspose2d(64, 32, 7),
            nn.ReLU(),
            nn.ConvTranspose2d(32, 16, 3, stride=2, padding=1, output_padding=1),
            nn.ReLU(),
            nn.ConvTranspose2d(16, nc, 3, stride=2, padding=1, output_padding=1),
            nn.Sigmoid()
        )
        self.emb = nn.Sequential(
            nn.Flatten(),
            nn.Linear(64 * 64 * nc, emb_dim)
        )

    def f(self, x):
        x = self.encoder(x)
        x = self.decoder(x)
        return x
    
    def forward(self, x):
        x = self.encoder(x)
        x = self.decoder(x)
        return x
    

basic_model = BasicModel(channels, emb_dim=5).to(device)
outer_opt = optim.Adam(basic_model.emb.parameters(), lr=outer_lr)
outer_mse_crit = nn.MSELoss()


for epoch in range(2):  # arbitrary number of epochs
    for batch_num in range(2):  # arbitrary number of outer steps
        frame = next(dataloader)

        inner_crit = nn.MSELoss()
        inner_opt = optim.SGD(list(basic_model.encoder.parameters()) +
                              list(basic_model.decoder.parameters()), lr=inner_lr)
        
        with higher.innerloop_ctx(
            basic_model,
            inner_opt,
            track_higher_grads=True,
            copy_initial_weights=False,
        ) as (fmodel, diffopt):
            for inner_step in range(2):  # arbitrary number of inner steps
                
                # NOTE: These two ways of computing a forward pass behave as desired,
                #       producing a gradient tape from `outer_loss` to emb params
                # gen_frame = fmodel(frame)
                # gen_frame = fmodel.forward(frame)
                
                # NOTE: There is no gradient path from `outer_loss` to embedding params
                #       when forward pass is computed in either of the following ways:
                gen_frame = fmodel.decoder(fmodel.encoder(frame))
                # seems like only the `forward` method gets special attention
                # gen_frame = fmodel.f(frame)

                embedding = fmodel.emb(gen_frame)
                # arbitrary simple stand-in loss to reproduce behavior
                inner_loss = inner_crit(torch.zeros_like(embedding), embedding)
                diffopt.step(inner_loss)

            # NOTE: These two ways of computing a forward pass behave as desired,
            #       producing a gradient tape from `outer_loss` to emb params
            # final_gen_frame = fmodel(frame)
            # final_gen_frame = fmodel.forward(frame)
            
            # NOTE: There is no gradient path from `outer_loss` to embedding params
            #       when forward pass is computed in either of the following ways:
            final_gen_frame = fmodel.decoder(fmodel.encoder(frame))
            # seems like only the `forward` method gets special attention
            # final_gen_frame = fmodel.f(frame)

        # arbitrary simple stand-in loss to reproduce behavior
        outer_loss = outer_mse_crit(torch.ones_like(final_gen_frame), final_gen_frame)

        # print gradients of `outer_loss` w.r.t. embedding parameters;
        #        `None` indicates no path
        dl_demb = torch.autograd.grad(outer_loss, basic_model.emb.parameters(),
                                      allow_unused=True, retain_graph=1)
        print(f'd(outer_loss)/d(emb): {dl_demb}')

        outer_opt.zero_grad()
        outer_loss.backward()
        outer_opt.step()
```
